File: app/src/main/java/com/example/scientific_calculator_app/MainActivity.java
package com.example.scientific_calculator_app;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.text.SpannableStringBuilder;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;
import org.mariuszgromada.math.mxparser.*;

public class MainActivity extends AppCompatActivity {

    private TextView previousCalculator;
    private EditText display;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        previousCalculator = findViewById(R.id.previousCalculationView);
        display = findViewById(R.id.displayEditText);
        display.setShowSoftInputOnFocus(false);
    }
    private void updateText(String strToAdd){
        String oldStr = display.getText ().toString();
        int cursorPos = display.getSelectionStart();
        String leftStr = oldStr.substring(0, cursorPos);
        String rightStr = oldStr.substring(cursorPos);

        display.setText(String.format("%s%s%s", oldStr, leftStr, strToAdd, rightStr));
        display.setSelection(cursorPos + strToAdd.length());
    }
    public void zeroBTNPush(View view){
        updateText(getResources().getString(R.string.zeroText));
    }
    public void oneBTNPush(View view){
        updateText(getResources().getString(R.string.oneText));
    }
    public void twoBTNPush(View view){
        updateText(getResources().getString(R.string.twoText));
    }
    public void threeBTNPush(View view){
        updateText(getResources().getString(R.string.threeText));
    }
    public void fourBTNPush(View view){
        updateText(getResources().getString(R.string.fourText));
    }
    public void fiveBTNPush(View view){
        updateText(getResources().getString(R.string.fiveText));
    }
    public void sixBTNPush(View view){
        updateText(getResources().getString(R.string.sixText));
    }
    public void sevenBTNPush(View view){
        updateText(getResources().getString(R.string.sevenText));
    }
    public void eightBTNPush(View view){
        updateText(getResources().getString(R.string.eightText));
    }
    public void nineBTNPush(View view){
        updateText(getResources().getString(R.string.nineText));
    }
    public void multiplyBTNPush(View view){
        updateText(getResources().getString(R.string.multiplyText));
    }
    public void dividePush(View view){
        updateText(getResources().getString(R.string.divideText));
    }
    public void subtractBTNPush(View view){
        updateText(getResources().getString(R.string.subtractText));
    }
    public void addBTNPush(View view){
        updateText(getResources().getString(R.string.addText));
    }

    public void clearBTNPush(View view){
        display.setText("");
        previousCalculator.setText("");

    }
    public void parOpenBTNPush(View view){
        updateText(getResources().getString(R.string.parenthesesOpenText));
    }
    public void parCloseBTNPush(View view){
        updateText(getResources().getString(R.string.parenthesesCloseText));
    }
    public void decimalBTNPush(View view){
        updateText(getResources().getString(R.string.decimalText));
    }
    public void equalBTNPush(View view){
        String userExp = display.getText().toString();

        previousCalculator.setText(userExp);

        userExp = userExp.replaceAll(getResources().getString(R.string.divideText), "/");
        userExp = userExp.replaceAll(getResources().getString(R.string.multiplyText), "*");

        Expression exp = new Expression(userExp);
        String result = String.valueOf(exp.calculate());

        display.setText(result);
        display.setSelection(result.length());

    }
    public void backspaceBTNPush(View view){
        int cursorPos = display.getSelectionStart();
        int textLen = display.getText ().length();

        if(cursorPos != 0 && textLen !=0){
            SpannableStringBuilder selection = (SpannableStringBuilder) display.getText();
            selection.replace(cursorPos-1, cursorPos, ";" );
            display.setText(selection);
            display.setSelection(cursorPos-1);

        }
    }

    public void trigSinBTNPush(View view){
        updateText("sin(");

    }
    public void trigCosBTNPush(View view){
        updateText("cos(");
    }
    public void trigTanBTNPush(View view){
        updateText("tan(");
    }
    public void trigArcSinBTNPush(View view){
        updateText("arcsin(");
    }
    public void trigArcCosBTNPush(View view){
        updateText("arccos(");
    }
    public void trigArcTanBTNPush(View view){
        updateText("arctan(");
    }
    public void naturalLogBTNPush(View view){
        updateText("ln(");
    }
    public void LogBTNPush(View view){
        updateText("log(");
    }
    public void squareRootBTNPush(View view){
        updateText("sqrt(");
    }
    public void absoluteValueBTNPush(View view){
        updateText("abs(");
    }
    public void piBTNPush(View view){
        updateText("pi");
    }
    public void eBTNPush(View view){
        updateText("e");
    }
    public void xSquaredBTNPush(View view){
        updateText("^(2)");
    }
    public void xPowerYBTNPush(View view){
        updateText("^(");
    }
    public void isPrimeFunctionBTNPush(View view){
        updateText("ispr(");
    }

}

File: app/src/main/java/com/example/scientific_calculator_app/MainActivity.kt
package com.example.scientific_calculator_app

import androidx.appcompat.app.AppCompatActivity
import android.widget.TextView
import android.widget.EditText
import android.os.Bundle
import com.example.scientific_calculator_app.R
import android.text.SpannableStringBuilder
import android.view.View
import org.mariuszgromada.math.mxparser.Expression

class MainActivity : AppCompatActivity() {
    private var previousCalculator: TextView? = null
    private var display: EditText? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        previousCalculator = findViewById(R.id.previousCalculationView)
        display = findViewById(R.id.displayEditText)
        display!!.setShowSoftInputOnFocus(false)
    }

    private fun updateText(strToAdd: String) {
        val oldStr = display!!.text.toString()
        val cursorPos = display!!.selectionStart
        val leftStr = oldStr.substring(0, cursorPos)
        val rightStr = oldStr.substring(cursorPos)
        display!!.setText(String.format("%s%s%s", leftStr, strToAdd, rightStr))
        display!!.setSelection(cursorPos + strToAdd.length)
    }

    fun zeroBTNPush(view: View?) {
        updateText(resources.getString(R.string.zeroText))
    }

    fun oneBTNPush(view: View?) {
        updateText(resources.getString(R.string.oneText))
    }

    fun twoBTNPush(view: View?) {
        updateText(resources.getString(R.string.twoText))
    }

    fun threeBTNPush(view: View?) {
        updateText(resources.getString(R.string.threeText))
    }

    fun fourBTNPush(view: View?) {
        updateText(resources.getString(R.string.fourText))
    }

    fun fiveBTNPush(view: View?) {
        updateText(resources.getString(R.string.fiveText))
    }

    fun sixBTNPush(view: View?) {
        updateText(resources.getString(R.string.sixText))
    }

    fun sevenBTNPush(view: View?) {
        updateText(resources.getString(R.string.sevenText))
    }

    fun eightBTNPush(view: View?) {
        updateText(resources.getString(R.string.eightText))
    }

    fun nineBTNPush(view: View?) {
        updateText(resources.getString(R.string.nineText))
    }

    fun multiplyBTNPush(view: View?) {
        updateText(resources.getString(R.string.multiplyText))
    }

    fun dividePush(view: View?) {
        updateText(resources.getString(R.string.divideText))
    }

    fun subtractBTNPush(view: View?) {
        updateText(resources.getString(R.string.subtractText))
    }

    fun addBTNPush(view: View?) {
        updateText(resources.getString(R.string.addText))
    }

    fun clearBTNPush(view: View?) {
        display!!.setText("")
        previousCalculator!!.text = ""
    }

    fun parOpenBTNPush(view: View?) {
        updateText(resources.getString(R.string.parenthesesOpenText))
    }

    fun parCloseBTNPush(view: View?) {
        updateText(resources.getString(R.string.parenthesesCloseText))
    }

    fun decimalBTNPush(view: View?) {
        updateText(resources.getString(R.string.decimalText))
    }

    fun equalBTNPush(view: View?) {
        var userExp = display!!.text.toString()
        previousCalculator!!.text = userExp
        userExp = userExp.replace(resources.getString(R.string.divideText).toRegex(), "/")
        userExp = userExp.replace(resources.getString(R.string.multiplyText).toRegex(), "*")
        val exp = Expression(userExp)
        val result = exp.calculate().toString()
        display!!.setText(result)
        display!!.setSelection(result.length)
    }

    fun backspaceBTNPush(view: View?) {
        val cursorPos = display!!.selectionStart
        val textLen = display!!.text.length
        if (cursorPos != 0 && textLen != 0) {
            val selection = display!!.text as SpannableStringBuilder
            selection.replace(cursorPos - 1, cursorPos, "")
            display!!.text = selection
            display!!.setSelection(cursorPos - 1)
        }
    }

    fun trigSinBTNPush(view: View?) {
        updateText("sin(")
    }

    fun trigCosBTNPush(view: View?) {
        updateText("cos(")
    }

    fun trigTanBTNPush(view: View?) {
        updateText("tan(")
    }

    fun trigArcSinBTNPush(view: View?) {
        updateText("arcsin(")
    }

    fun trigArcCosBTNPush(view: View?) {
        updateText("arccos(")
    }

    fun trigArcTanBTNPush(view: View?) {
        updateText("arctan(")
    }

    fun naturalLogBTNPush(view: View?) {
        updateText("ln(")
    }

    fun LogBTNPush(view: View?) {
        updateText("log(")
    }

    fun squareRootBTNPush(view: View?) {
        updateText("sqrt(")
    }

    fun absoluteValueBTNPush(view: View?) {
        updateText("abs(")
    }

    fun piBTNPush(view: View?) {
        updateText("pi")
    }

    fun eBTNPush(view: View?) {
        updateText("e")
    }

    fun xSquaredBTNPush(view: View?) {
        updateText("^(2)")
    }

    fun xPowerYBTNPush(view: View?) {
        updateText("^(")
    }

    fun isPrimeFunctionBTNPush(view: View?) {
        updateText("ispr(")
    }
}